File: portal/src/lib/api/grouping.ts
import { apiClient } from "./client";

export interface DeviceGroup {
  id: string;
  name: string;
  description: string | null;
  colour: string | null;
  device_count: number;
}

export interface UserTeam {
  id: string;
  name: string;
  description: string | null;
  colour: string | null;
  member_count: number;
}

export interface GroupWrite {
  name: string;
  description?: string | null;
  colour?: string | null;
}

export const listDeviceGroups = () =>
  apiClient.get<DeviceGroup[]>("/grouping/groups").then((r) => r.data);

export const createDeviceGroup = (body: GroupWrite) =>
  apiClient.post<DeviceGroup>("/grouping/groups", body).then((r) => r.data);

export const updateDeviceGroup = (id: string, body: GroupWrite) =>
  apiClient.patch<DeviceGroup>(`/grouping/groups/${id}`, body).then((r) => r.data);

export const deleteDeviceGroup = (id: string) =>
  apiClient.delete(`/grouping/groups/${id}`).then(() => undefined);

export const getGroupDevices = (id: string) =>
  apiClient.get<string[]>(`/grouping/groups/${id}/devices`).then((r) => r.data);

/** Replaces the group's devices with exactly this set — see MembershipWrite on the backend
 *  for why membership is sent whole rather than as add/remove deltas. */
export const setGroupDevices = (id: string, deviceIds: string[]) =>
  apiClient
    .put<DeviceGroup>(`/grouping/groups/${id}/devices`, { device_ids: deviceIds })
    .then((r) => r.data);

export interface GroupActionResult {
  action_id: string;
  /** "devices" or "sessions" — a session action fans out per signed-in session, so a
   *  terminal server counts once per person on it, not once as a machine. */
  fanned_over: string;
  targets: number;
  queued: number;
  failed: number;
  already_running: number;
  error: string | null;
}

/**
 * Push one action to everything in a group.
 *
 * Expert-plan only, same as the fleet-wide push — a 402 back means the plan, not the role.
 * Tiers are still enforced per device, so a technician aiming an admin-only action at 200
 * machines gets 200 refusals in `failed` rather than 200 sign-outs.
 */
export const runGroupAction = (
  groupId: string,
  body: { action_id: string; params?: Record<string, string>; message?: string; reason?: string },
) =>
  apiClient
    .post<GroupActionResult>(`/grouping/groups/${groupId}/actions`, body)
    .then((r) => r.data);

export const listUserTeams = () =>
  apiClient.get<UserTeam[]>("/grouping/teams").then((r) => r.data);

export const createUserTeam = (body: GroupWrite) =>
  apiClient.post<UserTeam>("/grouping/teams", body).then((r) => r.data);

export const updateUserTeam = (id: string, body: GroupWrite) =>
  apiClient.patch<UserTeam>(`/grouping/teams/${id}`, body).then((r) => r.data);

export const deleteUserTeam = (id: string) =>
  apiClient.delete(`/grouping/teams/${id}`).then(() => undefined);

export const getTeamUsers = (id: string) =>
  apiClient.get<string[]>(`/grouping/teams/${id}/users`).then((r) => r.data);

export const setTeamUsers = (id: string, userIds: string[]) =>
  apiClient
    .put<UserTeam>(`/grouping/teams/${id}/users`, { user_ids: userIds })
    .then((r) => r.data);

File: backend/app/schemas/grouping.py
import uuid

from pydantic import BaseModel, Field

# A 3/6/8-digit hex colour. Validated at the edge rather than trusted, because this value is
# written straight into a style attribute in the portal — a string that is not a colour is
# the beginning of a CSS injection, and there is no reason to accept one.
COLOUR_PATTERN = r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6}|[0-9a-fA-F]{8})$"


class GroupWrite(BaseModel):
    """Create/update body, shared by device groups and user teams — the two carry the same
    three fields, and giving them separate identical models only creates a chance for the
    two to drift apart later."""
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    colour: str | None = Field(default=None, pattern=COLOUR_PATTERN)


class DeviceGroupRead(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    colour: str | None
    # Counted server-side. The alternative is shipping every member id so the client can
    # take its length, which turns a list of 12 groups into a list of 2,000 device ids.
    device_count: int


class UserTeamRead(BaseModel):
    id: uuid.UUID
    name: str
    description: str | None
    colour: str | None
    member_count: int


class MembershipWrite(BaseModel):
    """Set a group's membership.

    The whole set, not a delta. Add/remove deltas race: two operators editing the same group
    in two tabs each send "add X", and neither can tell that the other removed Y. Sending
    the set the operator was looking at makes the last write win visibly, which is the
    behaviour they already expect from a list of checkboxes.
    """
    device_ids: list[uuid.UUID] = Field(default_factory=list, max_length=5000)


class TeamMembershipWrite(BaseModel):
    user_ids: list[uuid.UUID] = Field(default_factory=list, max_length=5000)


class GroupActionRequest(BaseModel):
    """One action, pushed to everything in a group.

    Deliberately the same shape whether it fans out over DEVICES (restart a service, install
    updates) or over SESSIONS (lock, sign out, message). The caller should not have to know
    which kind it picked — the service knows, because the registry knows, and making the
    client route it would be a second place for the two to disagree.
    """
    action_id: str = Field(min_length=1, max_length=60)
    # Free parameters the action declares — an app name to uninstall, a KB to install. Each
    # is still validated per-action by RemediationService; nothing here is trusted.
    params: dict[str, str] | None = None
    # message_session only.
    message: str | None = Field(default=None, max_length=1000)
    reason: str | None = Field(default=None, max_length=500)


class GroupActionResult(BaseModel):
    """What a bulk push actually did.

    `already_running` is separate from `failed` for the same reason it is on the fleet
    result: nothing went wrong on those machines, the work is already under way, and calling
    it a failure invites a second push that only duplicates it.

    `targets` is what the group resolved to at the moment of pushing — devices for a device
    action, live sessions for a session action. Reported back because a group's membership
    and a machine's session list both move, and "queued 12" means something different
    depending on whether the group held 12 devices or 400.
    """
    action_id: str
    fanned_over: str            # "devices" | "sessions"
    targets: int
    queued: int
    failed: int
    already_running: int = 0
    error: str | None = None
